File: src/Runtime/Module.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {

	public class Module {
		private string _name = "";
		private Executer _executer = null;
		private Space _space = null;
		private bool _bExecuted = false;

		public string name { get => _name; }
		public Space space { get => _space; }

		public Module(string name, Executer executer) {
			_name = name;
			_executer = executer;
			_space = _executer.space;
		}
		public Module(string name, Space space) {
			_name = name;
			_space = space;
			_bExecuted = true;
		}

		public bool Execute() {
			_bExecuted = true;
			if (_executer != null) {
				return _executer.Execute();
			}
			return true;
		}

		public bool IsExecuted() {
			return _bExecuted;
		}
	}

} // namespace peak.interpreter

File: src/runtime/module.cpp
#include "module.h"
#include "executer.h"

using namespace peak;

Module::Module(std::shared_ptr<Executer> executer)
	: _executer(executer) {
	_space = _executer->GetSpace();
}
Module::Module(std::shared_ptr<Space> space)
	: _space(space),
	_bExecuted(true) {
}

bool Module::Execute() {
	_bExecuted = true;
	if (_executer) {
		return _executer->Execute();
	}
	return true;
}

bool Module::IsExecuted() const {
	return _bExecuted;
}

std::shared_ptr<Space> Module::GetSpace() const {
	return _space;
}
